### PR TITLE
Implements Feedback for scheduled interviews

### PR DESCRIPTION
**Pull request by GrantComm**
Implements the feedback section and feedback forms. Will be merged after #116 
### The display on the scheduledInterview page: 
Screenshot 2020-07-29 at 10 09 19 AM

### Example of a feedback form: 
Screenshot 2020-07-29 at 10 11 27 AM

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/FeedbackServlet.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  @Override
+  public void init() {
+    init(new DatastoreScheduledInterviewDao());
+  }
+
+  public void init(ScheduledInterviewDao scheduledInterviewDao) {
+    this.scheduledInterviewDao = scheduledInterviewDao;
+  }
+
+  // Gets the current user's email from request and returns the ScheduledInterviews for that person.
```
Comment (meagancheese) [inline src/main/java/com/google/sps/servlets/FeedbackServlet.java]: Change this comment

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java [hunk 2]
```diff
@@ -62,6 +63,8 @@ public void init(ScheduledInterviewDao scheduledInterviewDao, PersonDao personDa
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String timeZoneId = request.getParameter("timeZone");
+    String userTime = request.getParameter("userTime");
+    System.out.println(userTime);
```
Comment (meagancheese) [inline src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Remove this print statement.

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java [hunk 3]
```diff
@@ -70,7 +73,8 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
       userId = String.format("%d", userEmail.hashCode());
     }
     List<ScheduledInterviewRequest> scheduledInterviews =
-        scheduledInterviewsToRequestObjects(scheduledInterviewDao.getForPerson(userId), timeZoneId);
+        scheduledInterviewsToRequestObjects(
+            scheduledInterviewDao.getForPerson(userId), timeZoneId, userTime);
```
Comment (meagancheese) [inline src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java:87]: Extra parantheses on the right of userId.

File: src/main/webapp/feedback.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+function getCurrentTime() {
+  return new Date().toISOString(); 
+}
+
+function getScheduledInterviewId() {
+  const queryString = window.location.search;
+  const urlParams = new URLSearchParams(queryString);
+  const interview = urlParams.get('interview')
```
Comment (meagancheese) [inline src/main/webapp/feedback.js]: Missing semi-colon.

File: src/main/webapp/feedback.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  console.log(interview);
+  return interview; 
+}
+
+function getRole() {
+  const queryString = window.location.search;
+  const urlParams = new URLSearchParams(queryString);
+  const role = urlParams.get('role')
```
Comment (meagancheese) [inline src/main/webapp/feedback.js]: Missing semi-colon.

File: src/main/webapp/feedback.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+function getCurrentTime() {
+  return new Date().toISOString(); 
+}
+
+function getScheduledInterviewId() {
+  const queryString = window.location.search;
+  const urlParams = new URLSearchParams(queryString);
+  const interview = urlParams.get('interview')
+  console.log(interview);
```
Comment (meagancheese) [inline src/main/webapp/feedback.js]: Remove this and add it back if you need to while debugging later.

File: src/main/webapp/feedback.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  return interview; 
+}
+
+function getRole() {
+  const queryString = window.location.search;
+  const urlParams = new URLSearchParams(queryString);
+  const role = urlParams.get('role')
+  console.log(role);
```
Comment (meagancheese) [inline src/main/webapp/feedback.js]: Remove this

File: src/main/webapp/feedback.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+function onFeedbackLoad() {
+  const loginInfo = getLoginInfo();
+  loginInfo.then(supplyLogoutLinkOrRedirectHome); 
+  loginInfo.then(getUserOrRedirectRegistration);
+  getScheduledInterviewId(); 
+  getRole();
```
Comment (meagancheese) [inline src/main/webapp/feedback.js]: These shouldn't be called here.

File: src/main/webapp/feedback.jsp [hunk 9]
```diff
@@ -0,0 +1,78 @@
+<%@ taglib uri="http://java.sun.com/jsp/jstl/core" prefix="c" %>
+<%
+  Boolean canOpen = (Boolean) request.getAttribute("feedbackOpen");
+  String role = (String) request.getAttribute("role");
+  pageContext.setAttribute("feedbackOpen", canOpen);
+  pageContext.setAttribute("role", role);
+  System.out.println(role); 
+  System.out.println(canOpen);
```
Comment (meagancheese) [inline src/main/webapp/feedback.jsp]: Remove these print statements.

File: src/main/webapp/feedback.jsp [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+          <label for="question1">The interviewee communicated their thought process as they went along:</label><br>
+          <input type="number" min="1" max="10" id="question1" name="question1"><br>
+          <label for="question2">The interviewee understood the time complexity of their solution:</label><br>
+          <input type="number" min="1" max="10" id="question2" name="question2"><br>
+          <label for="question3">The interviewee took the time to consider better solutions:</label><br>
+          <input type="number" min="1" max="10" id="question3" name="question3"><br>
+          <label for="question4">The interviewee came up with an example that they used to test their solution:</label><br>
+          <input type="number" min="1" max="10" id="question4" name="question4"><br>
+          <label for="question5">My interviewer answered my questions to the best of their ability:</label><br>
+          <input type="number" min="1" max="10" id="question5" name="question5"><br>
```
Comment (meagancheese) [inline src/main/webapp/feedback.jsp]: This needs to be removed (wrong section).

File: src/main/webapp/scheduled-interviews.jsp [hunk 11]
```diff
@@ -23,6 +23,14 @@
             <li class="list-group-item">Interviewee: ${scheduledInterview.getInterviewee()}</li>
             <li class="list-group-item">Interviewer: ${scheduledInterview.getInterviewer()}</li>
           </ul>
+          <c:choose>
+            <c:when test= "${scheduledInterview.getHasPassed()}">
+              <a href="feedback.html?interview=${scheduledInterview.getId()}&role=${scheduledInterview.getRole()}"><button class = "feedbackButton">Submit Feedback</button></a>
```
Comment (meagancheese) [inline src/main/webapp/scheduled-interviews.jsp]: Long line

File: src/main/webapp/scheduled-interviews.jsp [hunk 12]
```diff
@@ -23,6 +23,14 @@
...
+            </c:when>
+            <c:otherwise>
+              <p>You may submit feedback once the interview has commenced</p>
```
Comment (meagancheese) [inline src/main/webapp/scheduled-interviews.jsp]: Format this somehow, even if it's just centering it.

File: src/test/java/com/google/sps/servlets/FeedbackServletTest.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    MockHttpServletRequest getRequest = new MockHttpServletRequest();
+    MockHttpServletResponse getResponse = new MockHttpServletResponse();
+    helper.setEnvIsLoggedIn(true).setEnvEmail("user@company.org").setEnvAuthDomain("auth");
+    getRequest.addParameter("interview", "-1");
+    getRequest.addParameter("userTime", "2020-07-05T22:00:00Z");
+    getRequest.addParameter("timeZone", "Etc/UCT");
+    getRequest.addParameter("role", "Interviewer");
+    feedbackServlet.doGet(getRequest, getResponse);
+
+    Assert.assertEquals(false, (boolean) getRequest.getAttribute("feedbackOpen"));
```
Comment (meagancheese) [inline src/test/java/com/google/sps/servlets/FeedbackServletTest.java]: This does not match the comment for this test. Fix one of them.

**Review by meagancheese — CHANGES_REQUESTED** [review]
Some notes. Also, if you could do a formatting sweep, that would be great. Either centering or inlining or something with the questions and response boxes.

File: src/main/webapp/feedback.js (hunk 6 above)
Comment (GrantComm) [reply src/main/webapp/feedback.js]: Done

File: src/main/webapp/feedback.js (hunk 7 above)
Comment (GrantComm) [reply src/main/webapp/feedback.js]: Done

File: src/main/webapp/feedback.jsp (hunk 9 above)
Comment (GrantComm) [reply src/main/webapp/feedback.jsp]: Done

File: src/main/java/com/google/sps/servlets/FeedbackServlet.java (hunk 1 above)
Comment (GrantComm) [reply src/main/java/com/google/sps/servlets/FeedbackServlet.java]: Done

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java (hunk 2 above)
Comment (GrantComm) [reply src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Done

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java (hunk 3 above)
Comment (GrantComm) [reply src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java:87]: scheduledInterviewDao.getForPerson(userId) is being passed to the scheduledInterviewsToRequestObjects function

File: src/main/webapp/feedback.js (hunk 4 above)
Comment (GrantComm) [reply src/main/webapp/feedback.js]: Done

File: src/main/webapp/feedback.js (hunk 5 above)
Comment (GrantComm) [reply src/main/webapp/feedback.js]: Done

File: src/main/webapp/feedback.js (hunk 8 above)
Comment (GrantComm) [reply src/main/webapp/feedback.js]: Done

File: src/main/webapp/feedback.jsp (hunk 10 above)
Comment (GrantComm) [reply src/main/webapp/feedback.jsp]: Done

File: src/main/webapp/scheduled-interviews.jsp (hunk 12 above)
Comment (GrantComm) [reply src/main/webapp/scheduled-interviews.jsp]: Done

File: src/main/webapp/scheduled-interviews.jsp (hunk 11 above)
Comment (GrantComm) [reply src/main/webapp/scheduled-interviews.jsp]: Done

File: src/test/java/com/google/sps/servlets/FeedbackServletTest.java (hunk 13 above)
Comment (GrantComm) [reply src/test/java/com/google/sps/servlets/FeedbackServletTest.java]: Done

File: src/main/webapp/feedback.html [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+            <a class="nav-item nav-link" href="availability.html">Availability</a>
+            <a class="nav-item nav-link" href="search-interview.html">Scheduling</a>
+            <!--TODO: Move this to the right of the navbar-->
+            <a class="nav-item nav-link" href="#" id="login-tab">Logout</a>
+          </div>
+        </div>
+      </nav>
+      <br>
+      <div id="feedBackForm"> 
+      </div>
```
Comment (meagancheese) [inline src/main/webapp/feedback.html]: Nit: put </div> on same line.

File: src/main/webapp/feedback.jsp [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+<%
+  Boolean canOpen = (Boolean) request.getAttribute("feedbackOpen");
+  String role = (String) request.getAttribute("role");
+  pageContext.setAttribute("feedbackOpen", canOpen);
+  pageContext.setAttribute("role", role); 
+%>
+
+
+<c:choose>
+  <c:when test= "${canOpen == false}">
```
Comment (meagancheese) [inline src/main/webapp/feedback.jsp:10]: Nit: change to ! canOpen

File: src/main/webapp/scheduled-interviews.jsp [hunk 16]
```diff
@@ -23,6 +23,16 @@
...
             <li class="list-group-item">Interviewer: ${scheduledInterview.getInterviewer()}</li>
           </ul>
+          <c:choose>
+            <c:when test= "${scheduledInterview.getHasPassed()}">
+              <a href="feedback.html?interview=${scheduledInterview.getId()}&role=${scheduledInterview.getRole()}" style="text-align:center">
+                <button class = "feedbackButton">Submit Feedback</button>
+              </a>
+            </c:when>
+            <c:otherwise>
+              <p style="text-align:center">You may submit feedback once the interview has commenced</p>
```
Comment (meagancheese) [inline src/main/webapp/scheduled-interviews.jsp]: Nit: add period

**Review by meagancheese — APPROVED** [review]
Some nits, otherwise LGTM.

File: src/main/webapp/scheduled-interviews.jsp (hunk 16 above)
Comment (GrantComm) [reply src/main/webapp/scheduled-interviews.jsp]: Done

File: src/main/webapp/feedback.html (hunk 14 above)
Comment (GrantComm) [reply src/main/webapp/feedback.html]: Done

File: src/main/webapp/feedback.jsp (hunk 15 above)
Comment (GrantComm) [reply src/main/webapp/feedback.jsp:10]: When changed to !canOpen instead of canOpen == false, the conditional acts as if the value is true and displays the message when it should display the form

File: src/main/webapp/scheduled-interviews.jsp [hunk 17]
```diff
@@ -23,6 +23,16 @@
             <li class="list-group-item">Interviewee: ${scheduledInterview.getInterviewee()}</li>
             <li class="list-group-item">Interviewer: ${scheduledInterview.getInterviewer()}</li>
           </ul>
+          <c:choose>
+            <c:when test= "${scheduledInterview.getHasPassed()}">
+              <a href="feedback.html?interview=${scheduledInterview.getId()}&role=${scheduledInterview.getRole()}" style="text-align:center">
+                <button class = "feedbackButton">Submit Feedback</button>
```
Comment (apluscs) [inline src/main/webapp/scheduled-interviews.jsp]: Change the feedbackButton to a bootstrap style button to go along with the rest of the site

File: src/main/webapp/scheduled-interviews.jsp [hunk 18]
```diff
@@ -23,6 +23,16 @@
...
+              </a>
+            </c:when>
+            <c:otherwise>
+              <p style="text-align:center">You may submit feedback once the interview has commenced.</p>
```
Comment (apluscs) [inline src/main/webapp/scheduled-interviews.jsp]: Commenced l feel is so fancy, l had to look it up. Kudos on the good vocab but dumb it down please to "started" ?
Also just to be accurate change to 5 min before it starts

File: src/main/webapp/feedback.jsp [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+    <h2 style="text-align: center">You may not submit feedback yet.</h2>
+  </c:when>
+  <c:otherwise>
+    <c:choose>
+      <c:when test= "${role == 'Interviewee'}">
+        <h1 class="text-center">Please submit your feedback for your interviewer below</h1>
+        <!-- TODO: Send email on submit -->
+        <form action="/action_page.php" style="text-align:center">
```
Comment (apluscs) [inline src/main/webapp/feedback.jsp]: QQ: where does this action go?

File: src/main/webapp/feedback.jsp [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+          <h5>Please Enter a value between 1 and 10 (1 being strongly disagree and 10 being strongly agree)</h5>
+          <label for="question1">I was comfortable during the interview:</label><br>
```
Comment (apluscs) [inline src/main/webapp/feedback.jsp]: Personally l don't like how everything is centered. l think left align is more professional. Please change it. But if Meagan asked you to change it to center from left align l don't want to make you do even more work so we can leave it as center.

File: src/main/webapp/feedback.jsp [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+
+
+<c:choose>
+  <c:when test= "${canOpen == false}">
+    <h2 style="text-align: center">You may not submit feedback yet.</h2>
+  </c:when>
+  <c:otherwise>
+    <c:choose>
+      <c:when test= "${role == 'Interviewee'}">
+        <h1 class="text-center">Please submit your feedback for your interviewer below</h1>
```
Comment (apluscs) [inline src/main/webapp/feedback.jsp]: Titles should not be this long. Please just change it to "Feedback" and put the rest of the information in a subtitle

File: src/main/webapp/feedback.jsp [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+        <!-- TODO: Send email on submit -->
+        <form action="/action_page.php" style="text-align:center">
+          <h5>Please Enter a value between 1 and 10 (1 being strongly disagree and 10 being strongly agree)</h5>
```
Comment (apluscs) [inline src/main/webapp/feedback.jsp]: Nit: Enter --> enter

File: src/main/webapp/feedback.jsp (hunk 19 above)
Comment (apluscs) [inline src/main/webapp/feedback.jsp]: ls there anyway to make this a Bootstrap form to go along with everything else?

File: src/main/webapp/scheduled-interviews.js [hunk 23]
```diff
@@ -24,9 +24,13 @@ function getBrowserTimeZone() {
...
-  fetch(`/scheduled-interviews?timeZone=${getBrowserTimeZone()}`)
+  fetch(`/scheduled-interviews?timeZone=${getBrowserTimeZone()}&userTime=${getCurrentTime()}`)
     .then(response => response.text())
     .then(listOfCards => {
       document.getElementById('scheduled-interviews-cards').innerHTML = listOfCards;
     });
+    
+function getCurrentTime() {
+  return new Date().toISOString(); 
+}
```
Comment (apluscs) [inline src/main/webapp/scheduled-interviews.js]: Did something happen here? Brackets are messed up

File: src/main/java/com/google/sps/servlets/FeedbackServlet.java [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+    request.setAttribute("role", role);
+    RequestDispatcher rd = request.getRequestDispatcher("/feedback.jsp");
+    try {
+      rd.forward(request, response);
+    } catch (ServletException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private boolean feedbackIsOpen(
```
Comment (apluscs) [inline src/main/java/com/google/sps/servlets/FeedbackServlet.java:66]: Comment as to what this returns? So basically if it's 5 minutes before scheduledInterview starts?

**Review by apluscs — CHANGES_REQUESTED** [review]
Small things

File: src/main/webapp/feedback.jsp (hunk 20 above)
Comment (GrantComm) [reply src/main/webapp/feedback.jsp]: Yea Meagan suggested everything being centered

File: src/main/webapp/scheduled-interviews.jsp (hunk 17 above)
Comment (GrantComm) [reply src/main/webapp/scheduled-interviews.jsp]: Done

File: src/main/webapp/scheduled-interviews.jsp (hunk 18 above)
Comment (GrantComm) [reply src/main/webapp/scheduled-interviews.jsp]: Done

File: src/main/webapp/feedback.jsp (hunk 21 above)
Comment (GrantComm) [reply src/main/webapp/feedback.jsp]: Done

File: src/main/webapp/feedback.jsp (hunk 22 above)
Comment (GrantComm) [reply src/main/webapp/feedback.jsp]: Done

File: src/main/webapp/scheduled-interviews.js (hunk 23 above)
Comment (GrantComm) [reply src/main/webapp/scheduled-interviews.js]: Done

File: src/main/java/com/google/sps/servlets/FeedbackServlet.java (hunk 24 above)
Comment (GrantComm) [reply src/main/java/com/google/sps/servlets/FeedbackServlet.java:66]: Done

File: src/main/webapp/feedback.jsp (hunk 19 above)
Comment (GrantComm) [reply src/main/webapp/feedback.jsp]: At the moment it does nothing, email of the form would be in a separate pr

File: src/main/webapp/feedback.jsp (hunk 20 above)
Comment (meagancheese) [reply src/main/webapp/feedback.jsp]: It looked very bulky to me, seeing all the questions and input boxes to the left. I'd be open to maybe left aligned questions and right aligned inputs (for rating ones) or something like that. If all left-align is what you want, then that's fine too.

Comment (meagancheese) [reply src/main/webapp/feedback.jsp]: I still like the centering of the buttons and no access messages though. Mainly to be consistent with everywhere else on the site.

Comment (GrantComm) [reply src/main/webapp/feedback.jsp]: After discussion having form contents to the left and buttons centered is best

File: src/main/webapp/feedback.jsp (hunk 19 above)
Comment (GrantComm) [reply src/main/webapp/feedback.jsp]: Done

Comment (apluscs) [reply src/main/webapp/feedback.jsp]: We don't have an action_page.php right? So remove the action attribute for now?

File: src/main/webapp/feedback.jsp (hunk 20 above)
Comment (apluscs) [reply src/main/webapp/feedback.jsp]: l like centered buttons. Right aligned has the issue of varying question lengths. l think the input box on the next line would be better.
lt's also more work to validate the inputs if they can just type whatever. Maybe a dropdown with fields 1-10 would be better.

File: src/main/webapp/feedback.js [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+  const interview = urlParams.get('interview');
+  return interview; 
+}
+
+function getRole() {
+  const queryString = window.location.search;
+  const urlParams = new URLSearchParams(queryString);
+  const role = urlParams.get('role');
+  return role; 
+}
```
Comment (apluscs) [inline src/main/webapp/feedback.js]: Can we get some comments?

File: src/main/webapp/feedback.js [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+  const urlParams = new URLSearchParams(queryString);
+  const interview = urlParams.get('interview');
+  return interview; 
+}
+
+function getRole() {
+  const queryString = window.location.search;
+  const urlParams = new URLSearchParams(queryString);
+  const role = urlParams.get('role');
+  return role;
```
Comment (apluscs) [inline src/main/webapp/feedback.js]: You can inline these. Same for getScheduledInterviewId()

File: src/main/webapp/feedback.jsp (hunk 20 above)
Comment (GrantComm) [reply src/main/webapp/feedback.jsp]: Right now the scaled questions only accept numbers 1-10, no other values. An error message is shown when a number other than 1-10 is entered

File: src/main/java/com/google/sps/servlets/FeedbackServlet.java [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+    try {
+      rd.forward(request, response);
+    } catch (ServletException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  // Returns whether or not a scheduledInterviewId is found and the start time is at least five
+  // minutes
+  // before the user's current time
```
Comment (apluscs) [inline src/main/java/com/google/sps/servlets/FeedbackServlet.java]: Nit: fix this new-lining

File: src/main/webapp/feedback.css [hunk 28]
```diff
@@ -0,0 +1,15 @@
+#content {
+  margin-left: auto;
+  margin-right: auto;
+  width: 650px;
+}
+
+input {
+  width: 10% !important;
```
Comment (apluscs) [inline src/main/webapp/feedback.css:8]: Q: what does the `!important; ` do?

File: src/main/webapp/feedback.jsp [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+
+<c:choose>
+  <c:when test= "${canOpen == false}">
+    <h2 style="text-align: center">You may not submit feedback yet.</h2>
+  </c:when>
+  <c:otherwise>
+    <c:choose>
+      <c:when test= "${role == 'Interviewee'}">
+        <h1 class="text-center">Feedback</h1>
+        <h4 class="text-center">Please submit your feedback for your interviewer below</h4>
```
Comment (apluscs) [inline src/main/webapp/feedback.jsp]: Not a fan of how this is so big compared to the title, same for the h5 you have in the form. Please also get these 2 subtitles in the same box.
Can you style it like the top of this? https://interview-me-step-2020.appspot.com/availability.html

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java [hunk 30]
```diff
@@ -155,9 +160,12 @@ private ScheduledInterviewRequest makeScheduledInterviewRequest(
             .get(scheduledInterview.intervieweeId())
             .map(Person::firstName)
             .orElse("Nonexistent User");
+    String role = getUserRole(scheduledInterview, userId);
+    boolean hasPassed =
```
Comment (apluscs) [inline src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: l feel like this name is misleading. lt's not whether the interview has passed, it's more of whether it has started. Please change accordingly.

**Review by apluscs — CHANGES_REQUESTED** [review]
More small things.

File: src/main/webapp/scheduled-interviews.jsp [hunk 31]
```diff
@@ -23,6 +23,16 @@
             <li class="list-group-item">Interviewee: ${scheduledInterview.getInterviewee()}</li>
             <li class="list-group-item">Interviewer: ${scheduledInterview.getInterviewer()}</li>
           </ul>
+          <c:choose>
+            <c:when test= "${scheduledInterview.getHasPassed()}">
+              <a href="feedback.html?interview=${scheduledInterview.getId()}&role=${scheduledInterview.getRole()}" style="text-align:center">
+                <button class="btn btn-primary" type="button">Submit Feedback</button>
```
Comment (apluscs) [inline src/main/webapp/scheduled-interviews.jsp:29]: Can we disable this button when the user isn't allowed to access the form?

File: src/main/webapp/feedback.css (hunk 28 above)
Comment (GrantComm) [reply src/main/webapp/feedback.css:8]: It allows me to override the bootstrap css for the input field

File: src/main/webapp/feedback.jsp (hunk 20 above)
Comment (meagancheese) [reply src/main/webapp/feedback.jsp]: Input on same line as question / prompt looks much better. Thanks for working it out!

File: src/main/webapp/scheduled-interviews.jsp (hunk 31 above)
Comment (meagancheese) [reply src/main/webapp/scheduled-interviews.jsp:29]: When the user isn't allowed to access it, the form doesn't appear at all. Instead there's a "you don't have access yet" type message. Look in the jsp file for the specifics.

File: src/main/webapp/feedback.js (hunk 25 above)
Comment (GrantComm) [reply src/main/webapp/feedback.js]: Done

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java (hunk 30 above)
Comment (GrantComm) [reply src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Done

File: src/main/java/com/google/sps/servlets/FeedbackServlet.java (hunk 27 above)
Comment (GrantComm) [reply src/main/java/com/google/sps/servlets/FeedbackServlet.java]: Done

File: src/main/webapp/feedback.js (hunk 26 above)
Comment (GrantComm) [reply src/main/webapp/feedback.js]: Done

File: src/main/webapp/feedback.jsp (hunk 29 above)
Comment (GrantComm) [reply src/main/webapp/feedback.jsp]: Is the way I have it now better?

File: src/main/webapp/feedback.jsp [hunk 32]
```diff
@@ -0,0 +1,88 @@
...
+            <textarea rows="4" cols="50" class="form-control" id="question11" name="question11"></textarea><br>
+          </div>
+          <div style="text-align: center">
+            <button class="btn btn-primary" type="submit" style="text-align: center">Submit</button>
+          </div>
+        </form> 
+      </c:otherwise>
+    </c:choose>
+  </c:otherwise>
+</c:choose>
```
Comment (apluscs) [inline src/main/webapp/feedback.jsp]: Git complains about not having a new line at the end. Nit: please fix these wherever they occur (the red circle with a dash).

File: src/main/webapp/feedback.js [hunk 33]
```diff
@@ -0,0 +1,49 @@
...
+function getCurrentTime() {
+  return new Date().toISOString(); 
+}
+
+// Returns the id of the interview that feedback is for
+function getScheduledInterviewId() {
+  return new URLSearchParams(window.location.search).get('interview');  
+}
+
+// Returns the role a particular user
```
Comment (apluscs) [inline src/main/webapp/feedback.js]: Nit: role of
Nit: add periods to your comments (got called out on this a couple times).

File: src/main/webapp/feedback.jsp [hunk 34]
```diff
@@ -0,0 +1,88 @@
+<%@ taglib uri="http://java.sun.com/jsp/jstl/core" prefix="c" %>
+<%
+  boolean canOpen = (boolean) request.getAttribute("feedbackOpen");
+  String role = (String) request.getAttribute("role");
+  pageContext.setAttribute("feedbackOpen", canOpen);
+  pageContext.setAttribute("role", role); 
+%>
+
```
Comment (apluscs) [inline src/main/webapp/feedback.jsp:8]: Nit: rm this line

File: src/main/webapp/feedback.jsp (hunk 29 above)
Comment (apluscs) [reply src/main/webapp/feedback.jsp]: Yeah, just put periods at the ends of sentences please.

**Review by apluscs — CHANGES_REQUESTED** [review]
All nits

**Review by apluscs — APPROVED** [review]

File: src/main/webapp/feedback.jsp (hunk 32 above)
Comment (GrantComm) [reply src/main/webapp/feedback.jsp]: Done

File: src/main/webapp/feedback.js (hunk 33 above)
Comment (GrantComm) [reply src/main/webapp/feedback.js]: Done

File: src/main/java/com/google/sps/data/ScheduledInterviewRequest.java [hunk 35]
```diff
@@ -78,6 +90,8 @@ public String toString() {
         "interviewee",
         interviewee,
         "role",
-        role);
+        role,
+        "hasStarted",
```
Comment (acmcarther) [inline src/main/java/com/google/sps/data/ScheduledInterviewRequest.java]: Made this comment in another PR.

Does it make sense to inline the field names into the format string? I don't see any good reason to substitute them in.

File: src/main/java/com/google/sps/servlets/FeedbackServlet.java [hunk 36]
```diff
@@ -0,0 +1,77 @@
...
+  // Returns whether or not a scheduledInterviewId is found and the start time is at least five
+  // minutes before the user's current time.
+  private boolean feedbackIsOpen(
+      String userTimeString, String timeZoneString, Long scheduledInterviewId) {
+    ZoneId timeZoneId = ZoneId.of(timeZoneString);
+    Instant userTime = Instant.parse(userTimeString);
+    TimeRange scheduledInterviewRange =
+        scheduledInterviewDao.get(scheduledInterviewId).map(ScheduledInterview::when).orElse(null);
+    if (scheduledInterviewRange == null) {
+      throw new NullPointerException("This scheduledInterviewId does not exist");
```
Comment (acmcarther) [inline src/main/java/com/google/sps/servlets/FeedbackServlet.java:73]: This is sub-par user experience, since you're not relaying the error to frontend. 

I don't think there's enough time implement this right, but basically the JSP page should have a codepath for errors so that the page displays some obvious error and doesn't just appear to hang.

File: src/test/java/com/google/sps/servlets/FeedbackServletTest.java [hunk 37]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  // Tests that a NullPointerException is thrown when a scheduledInterviewId does not exist.
+  @Test(expected = NullPointerException.class)
+  public void interviewIdDoesNotExist() throws IOException {
```
Comment (acmcarther) [inline src/test/java/com/google/sps/servlets/FeedbackServletTest.java:75]: No test for the regular codepath?

File: src/main/webapp/feedback.jsp [hunk 38]
```diff
@@ -0,0 +1,87 @@
...
+      <c:when test= "${role == 'Interviewee'}">
+        <h1 class="text-center">Feedback</h1>
+        <h4 class="text-center">Please submit your feedback for your interviewer below.</h4>
+        <h5 class="text-center lead">Please enter a value between 1 and 10 (1 being strongly disagree and 10 being strongly agree).</h5>
+        <!-- TODO: Send email on submit -->
+        <form>
+          <div class="form-group">
+            <div class="form-group form-inline"> 
+              <label for="question1">I was comfortable during the interview:</label><br>
+              <input type="number" class="form-control" min="1" max="10" id="question1" name="question1"><br>
```
Comment (acmcarther) [inline src/main/webapp/feedback.jsp:24]: This might look better if these boxes are right-aligned.

(I don't feel strongly about this)

File: src/main/webapp/feedback.jsp [hunk 39]
```diff
@@ -0,0 +1,87 @@
...
+            <textarea rows="4" cols="50" class="form-control" id="question8" name="question8"></textarea><br>
+            <label for="question9">What is something that you think your interviewer could have done better?</label><br>
+            <textarea rows="4" cols="50" class="form-control" id="question9" name="question8"></textarea><br>
+          </div>
+          <div style="text-align: center">
+            <button class="btn btn-primary" type="submit" style="text-align: center">Submit</button>
+          </div>
+        </form>
+      </c:when>
+      <c:otherwise>
```
Comment (acmcarther) [inline src/main/webapp/feedback.jsp:60]: Explicitly test       `<c:when test= "${role == 'Interviewer'}">`

I've made this comment elsewhere. Things that are binary choices don't often stay binary choices, and when you structure your code this way where you assume `!one == the_other`, it makes adding a third option really onerous because each site where this was done needs to be amended, and if one is missed, logic will be broken unpredictably.

File: src/main/webapp/feedback.js [hunk 40]
```diff
@@ -0,0 +1,49 @@
...
+    }); 
+}
+
+// Returns the timezone of the user.
+function getBrowserTimeZone() {
+  return Intl.DateTimeFormat().resolvedOptions().timeZone; 
+}
+
+// Returns the user's current time.
+function getCurrentTime() {
```
Comment (acmcarther) [inline src/main/webapp/feedback.js]: Don't duplicate these functions. Instead, include them in a common file.

**Review by acmcarther — CHANGES_REQUESTED** [review]

File: src/main/webapp/feedback.js (hunk 40 above)
Comment (GrantComm) [reply src/main/webapp/feedback.js]: Done

File: src/main/java/com/google/sps/data/ScheduledInterviewRequest.java (hunk 35 above)
Comment (GrantComm) [reply src/main/java/com/google/sps/data/ScheduledInterviewRequest.java]: Done

File: src/main/java/com/google/sps/servlets/FeedbackServlet.java (hunk 36 above)
Comment (GrantComm) [reply src/main/java/com/google/sps/servlets/FeedbackServlet.java:73]: This is what is currently shown when the id is not found
Screenshot 2020-07-29 at 3 33 22 PM

File: src/main/webapp/feedback.jsp (hunk 39 above)
Comment (GrantComm) [reply src/main/webapp/feedback.jsp:60]: Done

File: src/main/java/com/google/sps/servlets/FeedbackServlet.java (hunk 36 above)
Comment (acmcarther) [reply src/main/java/com/google/sps/servlets/FeedbackServlet.java:73]: That's probably only in the development mode (else, that's a security risk), but it's better than I expecte.d

File: src/main/webapp/feedback.jsp (hunk 38 above)
Comment (GrantComm) [reply src/main/webapp/feedback.jsp:24]: Done

File: src/test/java/com/google/sps/servlets/FeedbackServletTest.java (hunk 37 above)
Comment (GrantComm) [reply src/test/java/com/google/sps/servlets/FeedbackServletTest.java:75]: Done

**Review by acmcarther — APPROVED** [review]
